pallets/router/src/lib.rs

```diff
@@ -0,0 +1,193 @@
+        TradedSuccessfully(T::AccountId, Balance, Route, Balance),
+    }
+
+    #[pallet::hooks]
+    impl<T: Config> Hooks<BlockNumberFor<T>> for Pallet<T> {}
+
+    #[pallet::call]
+    impl<T: Config> Pallet<T> {
+        /// According specified route order to execute which pool or AMM instance.
+        ///
+        /// - `origin`: the trader.
+        /// - `route`: the route user inputs
+        /// - `amount_in`: the amount of trading assets
+        /// - `min_amount_out`:
+        /// - `expiry`:
+        #[pallet::weight(10_000)]
+        #[transactional]
+        pub fn trade(
+            origin: OriginFor<T>,
+            route: Route,
+            #[pallet::compact] amount_in: Balance,
+            #[pallet::compact] min_amount_out: Balance,
+            #[pallet::compact] expiry: BlockNumberFor<T>,
+        ) -> DispatchResultWithPostInfo {
+            let trader = ensure_signed(origin)?;
+
+            // Ensure the length of routes should be >= 1 at least.
+            ensure!(!route.is_empty(), Error::<T>::EmptyRoute);
+            // Ensure user do not input too many routes.
+            ensure!(
+                route.len() <= T::MaxLengthRoute::get() as usize,
+                Error::<T>::ExceedMaxLengthRoute
+            );
+
+            // Ensure user doesn't input duplicated routes
+            let mut _routes = route.clone();
+            _routes.dedup();
+            ensure!(_routes == route, Error::<T>::DuplicatedRoute);
+
+            // Ensure balances user input is bigger than zero.
+            ensure!(
+                amount_in > Zero::zero() && min_amount_out >= Zero::zero(),
+                Error::<T>::ZeroBalance
+            );
+
+            // Ensure user iput a valid block number.
+            let current_block_num = <frame_system::Pallet<T>>::block_number();
+            ensure!(expiry > current_block_num, Error::<T>::TooSmallExpiry);
+
+            // Ensure the trader has enough tokens for transaction.
+            let (_, from_currency_id, _) = route[0];
+            ensure!(
+                T::Currency::free_balance(from_currency_id, &trader) > amount_in,
+                Error::<T>::InsufficientBalance
+            );
+
+            // Get all AMM routes we're supporting now.
+            let all_routes = T::Routes::get();
+
+            // Ensure the routes user input are valid.
+            ensure!(
+                route.iter().all(|r| all_routes.iter().any(|_r| r == _r)),
+                Error::<T>::NotSupportedRoute
+            );
+
+            let mut amount_out: Balance = Zero::zero();
+            for sub_route in route.iter() {
+                let (_id, from_currency_id, to_currency_id) = sub_route;
+
+                amount_out = amount_in;
+                amount_out = T::AMMAdaptor::trade(
+                    &trader,
+                    (*from_currency_id, *to_currency_id),
+                    amount_out,
+                    min_amount_out,
+                )?;
```

Comment:
```suggestion
                amount_out = T::AMMAdaptor::trade(
                    &trader,
                    (*from_currency_id, *to_currency_id),
                    amount_in,
                    min_amount_out,
                )?;
```

this got me confused 😉